weasyl/commishinfo.py

```diff
@@ -20,39 +53,261 @@ def convert_currency(target):
         return 0
     # strip everything but digits and a decimal point
     digits = re.sub("[^0-9.]", "", target)
+    if not digits:
+        return 0
     return int(Decimal(digits) * (10 ** CURRENCY_PRECISION))
 
 
-def select_list(userid):
-    query = d.execute("SELECT classid, title, amount_min, amount_max, settings, priceid FROM commishprice"
-                      " WHERE userid = %i ORDER BY classid, title", [userid])
+@region.cache_on_arguments(expiration_time=60 * 60 * 24)
+def _fetch_rates():
+    """
+    Retrieves most recent currency exchange rates from fixer.io, which uses
+    the European Central Bank as its upstream source.
+
+    This value is cached with a 24h expiry period
+
+    :return: see http://fixer.io/
+    """
+    try:
+        return d.http_get("http://api.fixer.io/latest?base=USD").json()
+    except WeasylError:
+        # There was an HTTP error while fetching from the API
+        return None
+
+
+def _charmap_to_currency_code(charmap):
+    """
+    Convert Weasyl's internal single-character representation of currencies
+    to standard ISO4217 codes for use in comparing against exchange rate APIs
+
+    :param charmap: String containing ideally one or zero characters used as currency indicators by Weasyl
+    :return: A 3-letter ISO4217 currency code. Returns "USD" if no match found.
+    """
+    for c in charmap:
+        if c in CURRENCY_CHARMAP:
+            return CURRENCY_CHARMAP.get(c)['code']
+    return "USD"
+
+
+def convert_currency(value, valuecode, targetcode):
+    """
+    Convert between different currencies using the current exchange rate.
+
+    :param value: The amount of currency to be converted
+    :param valuecode: Weasyl-code of the currency $value is in currently
+    :param targetcode: Weasyl-code of the currency $value should be converted to
+    :return: The converted amount
+    """
+    if targetcode == valuecode:
+        return value
+    ratio = currency_ratio(valuecode, targetcode)
+    return value * ratio
+
+
+def currency_ratio(valuecode, targetcode):
+    """
+    Calculate the exchange rate multiplier used to convert from $valuecode to $targetcode
+
+    :param valuecode: Weasyl-code of the currency $value is in currently
+    :param targetcode: Weasyl-code of the currency $value should be converted to
+    :return: The conversion ratio
+    """
+    valuecode = _charmap_to_currency_code(valuecode)
+    targetcode = _charmap_to_currency_code(targetcode)
+    rates = _fetch_rates()
+    if not rates:
+        # in the unlikely event of an error, invalidate our rates
+        # (to try fetching again) and return value unaltered
+        _fetch_rates.invalidate()
+        return 1.0
```

Comment:
You are probably right, but what *should* be the right value? Raising an exception seems excessive considering the conversion is more of a "nice to have". than "required functionality". None might make sense but it is slightly inconvenient and i'd probably have to assume a value for the sql query anyway. 0 is an obviously erroneous non-null value and I GUESS for truly robust code I should assume it's possible for fixer.io to send 0 or some other malformed value as well.

How it works right now is that it doesn't display the "converted" price if it's equal to the original price, covering both the cases of "they are the same currency" and "disastrous failure of the service we are using". But is that putting too many assumptions into the view? Please advise.